muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/guider/GuiderRESTService.java

```diff
@@ -1045,4 +1086,175 @@ else if (value instanceof Double) {
+      return Response.status(Status.NOT_FOUND).build();
+    }
+
+    Workspace workspace = workspaceController.findWorkspace(workspaceEntity);
+
+    SchoolDataIdentifier workspaceIdentifier = new SchoolDataIdentifier(workspace.getIdentifier(), workspace.getSchoolDataSource());
+
+    WorkspaceUserEntity workspaceUserEntity = workspaceUserEntityController.findWorkspaceUserEntityByWorkspaceAndUserIdentifierIncludeArchived(workspaceEntity, studentIdentifier);
+    if (workspaceUserEntity != null && workspaceUserEntity.getArchived() == Boolean.TRUE) {
+      workspaceUserEntityController.unarchiveWorkspaceUserEntity(workspaceUserEntity);
+    }
+    if (workspaceUserEntity != null && workspaceUserEntity.getActive() == Boolean.FALSE) {
+      workspaceUserEntityController.updateActive(workspaceUserEntity, Boolean.TRUE);
+      userIndexer.indexUser(workspaceUserEntity.getUserSchoolDataIdentifier().getUserEntity());
+    }
+
+    fi.otavanopisto.muikku.schooldata.entity.WorkspaceUser workspaceUser = workspaceController.findWorkspaceUserByWorkspaceAndUser(workspaceIdentifier, studentEntity.defaultSchoolDataIdentifier());
+    if (workspaceUser == null) {
+      workspaceUser = workspaceController.createWorkspaceUser(workspace, student, WorkspaceRoleArchetype.STUDENT);
+      waitForWorkspaceUserEntity(workspaceEntity, studentIdentifier);
+    }
+    else {
+      workspaceController.updateWorkspaceStudentActivity(workspaceUser, true);
+    }
+
+    List<WorkspaceUserEntity> workspaceTeachers = workspaceUserEntityController.listActiveWorkspaceStaffMembers(workspaceEntity);
+    List<UserEntity> recipients = new ArrayList<UserEntity>();
+
+    String workspaceName = workspace.getName();
+    if (!StringUtils.isBlank(workspace.getNameExtension())) {
+      workspaceName += String.format(" (%s)", workspace.getNameExtension()); 
+    }
+
+    String userName = student.getNickName() == null
+      ? student.getDisplayName()
+      : String.format("%s \"%s\" %s (%s)", student.getFirstName(), student.getNickName(), student.getLastName(), student.getStudyProgrammeName());
+
+    User loggedUser = userController.findUserByUserEntityDefaults(sessionController.getLoggedUserEntity());
+
+    String loggedUserName = loggedUser.getNickName() == null
+        ? loggedUser.getDisplayName()
+        : String.format("%s \"%s\" %s (%s)", loggedUser.getFirstName(), loggedUser.getNickName(), loggedUser.getLastName(), loggedUser.getStudyProgrammeName());
+
+    recipients.add(studentEntity);
+
+    for (WorkspaceUserEntity workspaceTeacher : workspaceTeachers) {
+      recipients.add(workspaceTeacher.getUserSchoolDataIdentifier().getUserEntity());
+    }
+
+    UserSchoolDataIdentifier userSchoolDataIdentifier = userSchoolDataIdentifierController.findUserSchoolDataIdentifierBySchoolDataIdentifier(studentIdentifier);
+
+    workspaceController.createWorkspaceUserSignup(workspaceEntity, userSchoolDataIdentifier.getUserEntity(), new Date(), entity.getMessage());
```

Comment:
Why `userSchoolDataIdentifier.getUserEntity()`? Didn't you already figure out `studentEntity` towards the very beginning of the method?